code/modules/jobs/job_types/security_officer.dm

```diff
@@ -149,11 +205,228 @@ GLOBAL_LIST_INIT(available_depts, list(SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, S
 	keyslot = new /obj/item/encryptionkey/headset_sec
 	keyslot2 = new /obj/item/encryptionkey/headset_sci
+
+/// Returns the distribution of splitting the given security officers into departments.
+/// Return value is an assoc list of candidate => SEC_DEPT_*.
+/proc/get_officer_departments(list/preferences, list/departments)
+	if (!preferences.len)
+		return list()
+
+	/**
+	 * This is a pretty complicated algorithm, but it's one I'm rather proud of.
+	 *
+	 * This is the function that is responsible for taking the list of preferences,
+	 * and spitting out what to put them in.
+	 *
+	 * However, it should, wherever possible, prevent solo departments.
+	 * That means that if there's one medical officer, and one engineering officer,
+	 * that they should be put onto the same department (either medical or engineering).
+	 *
+	 * The first step is to get the "distribution". This describes how many officers
+	 * should be in each department, no matter what they are.
+	 * This is handled in `get_distribution`. Examples of inputs/outputs are:
+	 * get_distribution(1, 4) => [1]
+	 * get_distribution(2, 4) => [2]
+	 * get_distribution(3, 4) => [3] # If this returned [2, 1], then we'd get a loner.
+	 * get_distribution(4, 4) => [2, 2] # We have enough to put into a separate group
+	 *
+	 * Once this distribution is received, the next step is to figure out where to put everyone.
+	 *
+	 * If all members have no preference, just make one an unused department (from the departments argument).
+	 * Then, call ourselves again.
+	 *
+	 * Order the groups from most populated to least.
+	 *
+	 * If the top group has enough officers who actually *want* that department, then we give it to them.
+	 * If there are any leftovers (for example, if 3 officers want medical, but we only want 2), then we
+	 * update those to have no preference instead.
+	 *
+	 * If the top group does NOT have enough officers, then we kill the least popular group by setting
+	 * them all to have no preference.
+	 *
+	 * Anyone in the most popular group will be removed from the list, and the final tally will be updated.
+	 * In the case of not having enough officers, this is a no-op, as there won't be any in the most popular group yet.
+	 *
+	 * If there are any candidates left, then we call the algorithm again, but for everyone who hasn't been selected yet.
+	 * We take the results from that run, and put them in the correct order.
+	 *
+	 * As an example, let's assume we have the following preferences:
+	 * [engineer, medical, medical, medical, medical, cargo]
+	 *
+	 * The distribution of this is [2, 2, 2], meaning there will be 3 departments chosen and they will have 2 each.
+	 * We order from most popular to least popular and get:
+	 * - medical: 4
+	 * - engineer: 1
+	 * - cargo: 1
+	 *
+	 * We need 2 to fill the first group. There are enough medical staff to do it. Thus, we take the first 2 medical staff
+	 * and update the output, making it now: [?, medical, medical, ?, ?, cargo].
```

Comment:
Why isn't engineer in this list? am I misunderstanding something?

Reply:
Typo, looks like.

```suggestion
	 * and update the output, making it now: [engineer, medical, medical, ?, ?, cargo].
```